_posts/2016-11-18-present-vs-any-vs-exists.markdown

```diff
@@ -0,0 +1,80 @@
+=> true
+```
+
+As you can see in the snippet above, we're loading one project, and then loading
+all of its tasks to check for presence using `present?`. This ends up taking
+quite a bit of time, hurting the performance of the app both memory and
+time-wise.
+
+A slightly better approach would be using `any?`:
+
+```ruby
+irb(main):004:0> Project.find(57).tasks.where.not(deleted_at: nil).any?
+D, [2016-11-14T19:44:57.288763 #27028] DEBUG -- :   Project Load (0.9ms)  SELECT  "projects".* FROM "projects"  WHERE "projects"."enabled" = 't' AND "projects"."id" = $1 LIMIT 1  [["id", 57]]
+D, [2016-11-14T19:44:57.410973 #27028] DEBUG -- :    (119.0ms)  SELECT COUNT(*) FROM "tasks" INNER JOIN "boards" ON "tasks"."board_id" = "boards"."id" WHERE "tasks"."enabled" = 't' AND "boards"."project_id" = $1 AND "boards"."enabled" = 't' AND ("tasks"."deleted_at" IS NOT NULL)  [["project_id", 57]]
+=> true
+```
+
+`any?` uses SQL count instead of loading each task, resulting in a faster, more
+performant result. However, what we actually want to know in this case is if
+there is at least one record in our scope. We don't really need to count all of
+the tasks, it should stop after finding the first one. So applying `LIMIT` would
+solve that for us, and that's how `exists?` saves the day:
+
+```ruby
+irb(main):005:0> Project.find(57).tasks.where.not(deleted_at: nil).exists?
+D, [2016-11-14T19:45:01.413654 #27028] DEBUG -- :   Project Load (0.5ms)  SELECT  "projects".* FROM "projects"  WHERE "projects"."enabled" = 't' AND "projects"."id" = $1 LIMIT 1  [["id", 57]]
+D, [2016-11-14T19:45:01.421724 #27028] DEBUG -- :   Task Exists (0.9ms)  SELECT  1 AS one FROM "tasks" INNER JOIN "boards" ON "tasks"."board_id" = "boards"."id" WHERE "tasks"."enabled" = 't' AND "boards"."project_id" = $1 AND "boards"."enabled" = 't' AND ("tasks"."deleted_at" IS NOT NULL) LIMIT 1  [["project_id", 57]]
+=> true
+```
+
+By limiting the count to 1, after finding the first record, it returns true.
+Notice the time in the debug lines in parenthesis, we went from ~900ms using
+`present?`, to ~100ms using `any?`, to ~1ms using `exists?`.
+
+Finally, here's a benchmark (using [benchmark-ips](https://github.com/evanphx/benchmark-ips)) which shows the difference in execution time between the three methods:
+
+```ruby
+Benchmark.ips do |x|
+  x.report("present?") do
+    Project.find(1).tasks.where.not(deleted_at: nil).present?
+  end
+  x.report("any?") do
+    Project.find(1).tasks.where.not(deleted_at: nil).any?
+  end
+  x.report("exists?") do
+    Project.find(1).tasks.where.not(deleted_at: nil).exists?
+  end
+  x.compare!
+end
+
+ exists?:      158.4 i/s
+    any?:       10.1 i/s - 15.65x  slower
+present?:        2.3 i/s - 68.39x  slower
+```
+
+NOTE: The examples shown in this post are taken from a Rails 4.1 app, using
+PostgreSQL. I am not sure if this behavior remains the same in newer versions
+of Rails.
```

Comment:
@mauro-oto Do the results remain so if you change the order of the blocks in the ips call? 

I'm asking because I wonder if there is a query cache in the middle that is affecting the results. 

Also, would it be too complicated to get results using Rails 4.2?